engine/test-utils/src/main/java/io/deephaven/engine/testutil/HierarchicalTableTestTools.java

```diff
@@ -0,0 +1,125 @@
+     * </p>
+     *
+     * @param hierarchicalTable the {@link HierarchicalTable} to snapshot
+     * @param snapshotState the state created from the {@link HierarchicalTable#makeSnapshotState()} on
+     *        hierarchicalTable
+     * @param keyTable the keyTable for expansions
+     * @param keyTableActionColumn the Action column within the key talbe
+     * @param columns the columns to snapshot (null for all of htem)
+     * @param rows the row positions to snapshot
+     * @return a Table containing the snapshot. Must be released with {@link #freeSnapshotTableChunks(Table)} after use.
+     */
+    @SuppressWarnings("SameParameterValue")
+    public static Table snapshotToTable(
+            @NotNull final HierarchicalTable<?> hierarchicalTable,
+            @NotNull final HierarchicalTable.SnapshotState snapshotState,
+            @NotNull final Table keyTable,
+            @Nullable final ColumnName keyTableActionColumn,
+            @Nullable final BitSet columns,
+            @NotNull final RowSequence rows) {
+        final ColumnDefinition<?>[] availableColumns =
+                hierarchicalTable.getAvailableColumnDefinitions().toArray(ColumnDefinition[]::new);
+        final ColumnDefinition<?>[] includedColumns = columns == null
+                ? availableColumns
+                : columns.stream().mapToObj(ci -> availableColumns[ci]).toArray(ColumnDefinition[]::new);
+
+        assertThat(rows.isContiguous()).isTrue();
+        final int rowsSize = rows.intSize();
+        // noinspection rawtypes
+        final WritableChunk[] chunks = Arrays.stream(includedColumns)
+                .map(cd -> maybeConvertToPrimitiveChunkType(cd.getDataType()))
+                .map(ct -> ct.makeWritableChunk(rowsSize))
+                .toArray(WritableChunk[]::new);
+
+        // noinspection unchecked
+        final long expandedSize =
+                hierarchicalTable.snapshot(snapshotState, keyTable, keyTableActionColumn, columns, rows, chunks);
+        final int snapshotSize = chunks.length == 0 ? 0 : chunks[0].size();
+        final long expectedSnapshotSize = rows.isEmpty()
+                ? 0
+                : Math.min(rows.lastRowKey() + 1, expandedSize) - rows.firstRowKey();
+        assertThat(snapshotSize).isEqualTo(expectedSnapshotSize);
+
+        final LinkedHashMap<String, ColumnSource<?>> sources = new LinkedHashMap<>(includedColumns.length);
+        for (int ci = 0; ci < includedColumns.length; ++ci) {
+            final ColumnDefinition<?> columnDefinition = includedColumns[ci];
+            // noinspection unchecked
+            final WritableChunk<? extends Values> chunk = chunks[ci];
+            final ChunkColumnSource<?> chunkColumnSource = ChunkColumnSource.make(
+                    chunk.getChunkType(), columnDefinition.getDataType(), columnDefinition.getComponentType());
+            if (snapshotSize > 0) {
+                chunkColumnSource.addChunk(chunk);
+            }
```

Comment:
Note, the check for > 0 is the only functional change here.

Reply:
Maybe we could just skip this complicated loop in that case?

Reply:
Decided I don't want to worry about that, because we need at least the part that is going to create the NVCS in the loop.